### PR TITLE
Implements the readdir recursive flag

### PR DESCRIPTION
**Pull request by arcanis**
**What's the problem this PR addresses?**
The `ZipFS` implementation doesn't support the `recursive` flag.

Fixes part of #5423

**How did you fix it?**
I implemented the `recursive` flag and updated the types to account for it. However I didn't make `ZipOpenFS` automatically traverse through zip archives ... My line of thinking was:

> - on one hand, tools implementing their own readdir-based recursive traversal don't currently dig into zip archives
> - on the other hand, it can be argued that the other recursive Node.js native operations also apply to the zip archives' content (`cp` / `rm`)
>
> But I tend to think the first point is more important ... I feel like the recursive flag  should just be an optimization, so it should have the same behaviour as a manual traversal ...

**Checklist**

- [x] I have read the [Contributing Guide](https://yarnpkg.com/advanced/contributing).

- [x] I have set the packages that need to be released for my changes to be effective.

- [x] I will check that all automated PR checks pass before the PR gets reviewed.

### CODE REVIEW

File: packages/yarnpkg-pnpify/sources/NodeModulesFS.ts [hunk 1]
```diff
@@ -500,53 +501,75 @@ export class PortableNodeModulesFS extends FakeFS<PortablePath> {
...
+  async readdirPromise(p: PortablePath, opts: {recursive: true, withFileTypes?: false}): Promise<Array<PortablePath>>;
+  async readdirPromise(p: PortablePath, opts: {recursive: true, withFileTypes: boolean}): Promise<Array<Dirent | PortablePath>>;
+  async readdirPromise(p: PortablePath, opts: {recursive: boolean, withFileTypes: true}): Promise<Array<Dirent<PortablePath>>>;
+  async readdirPromise(p: PortablePath, opts: {recursive: boolean, withFileTypes?: false}): Promise<Array<PortablePath>>;
+  async readdirPromise(p: PortablePath, opts: {recursive: boolean, withFileTypes: boolean}): Promise<Array<Dirent<PortablePath> | PortablePath>>;
+  async readdirPromise(p: PortablePath, opts?: ReaddirOptions | null): Promise<Array<Dirent<PortablePath> | Dirent | PortablePath>> {
     const pnpPath = this.resolvePath(p);
     if (pnpPath.dirList || this.resolvePath(ppath.join(p, `node_modules`)).dirList) {
+      if (opts?.recursive)
+        throw new Error(`Unsupported option 'recursive' for NodeModulesFS.readdirPromise`);
```
Comment (arcanis) [inline packages/yarnpkg-pnpify/sources/NodeModulesFS.ts:518]: I'm not sure how to implement it in this FakeFS

File: packages/plugin-pnpm/sources/PnpmLinker.ts [hunk 2]
```diff
@@ -85,7 +85,7 @@ class PnpmInstaller implements Installer {
   private readonly indexFolderPromise: Promise<PortablePath>;
 
   constructor(private opts: LinkOptions) {
-    this.indexFolderPromise = setupCopyIndex(xfs, {
+    this.indexFolderPromise = setupCopyIndex<PortablePath>(xfs, {
```
Comment (arcanis) [inline packages/plugin-pnpm/sources/PnpmLinker.ts:88]: I have no idea why, TypeScript stopped inferring the PortablePath generic 😕

File: packages/yarnpkg-pnpify/sources/NodeModulesFS.ts [hunk 3]
```diff
@@ -1,3 +1,4 @@
+import {ReaddirOptions}                                                                                                            from '@yarnpkg/fslib/sources/FakeFS';
```
Comment (paul-soporan) [inline packages/yarnpkg-pnpify/sources/NodeModulesFS.ts:1]:
```suggestion
import {ReaddirOptions}                                                                                                            from '@yarnpkg/fslib';
```

File: packages/yarnpkg-libzip/sources/ZipFS.ts [hunk 4]
```diff
@@ -1,3 +1,4 @@
+import {ReaddirOptions}                                                                                                                              from '@yarnpkg/fslib/sources/FakeFS';
```
Comment (paul-soporan) [inline packages/yarnpkg-libzip/sources/ZipFS.ts:1]:
```suggestion
import {ReaddirOptions}                                                                                                                              from '@yarnpkg/fslib';
```

**Pull request merged**